settlement-methods-location.md

```diff
@@ -0,0 +1,84 @@
+# Settlement methods for messages and impact on users
+
+In the current design for v7 of the Service Bus SDK, across the 4 supported languages, there
+are 3 places where settlement methods (complete, abandon, defer, deadLetter) live:
+
+- On the receiver (Java, C#)
+- On an object passed into the callback where you received your messages (C# via ServiceBusProcessor)
+- On the message itself (Python, JavaScript)
+
+There are some benefits to the JavaScript chosen path:
+
+- Developers will not be forced to pass in a receiver to code that only expects to use or settle a message.
+  This can aid with creating a better separation of concerns.
+- The message is a likely location for settlement methods (debatable, but we've landed on this side of the debate)
+- The current stable version of Azure Service Bus package for JavaScript has the settlement methods
+  on the message. Compatibility is a strong argument.
+- Settlement methods on the Receiver clutter up the interface.
+
+Up until recently, the chosen location was primarily cosmetic.
+
+## It's not cosmetic?
+
+One use case that has come up is this:
+
+```javascript
+receiver.subscribe({
+  processMessage: (msg) => {
+    // (fire and forget - we don't need this message in this process anymore)
+    const arbitraryTransformedData = arbitraryTransformForOtherService(
+      msg.body
+    );
+    sendMessageToSomeOtherService(arbitrarilyTransformedData, msg.lockToken);
+
+    // at this point the user wants to throw the contents of the message away
+    // to save memory, for instance.
+  },
+  // ...rest of code elided ...
+});
+```
+
+Which results in activity like this:
+
+1. `receiver` -> sendMessageToSomeOtherService(arbitraryData, lockToken) -> `some other service`
+2. `some other service` does some processing and now wants `receiver` to complete it.
+3. `some other service` sends back the lock token (no need to send the _entire_ message or even parts of it)
+4. `receiver` can complete with just the lock token.
+
+Our currently chosen location for message completion doesn't provide a simple way to accomplish this task
+using _only_ the lock token.
+
+## Ideas
+
+### Provide additional lock token overloads
+
+We could add additional completion methods to Receiver.
+
+```typescript
+interface Receiver {
+  complete(lockTokenOrMessage: string | { lockToken: string }, options);
+  abandon(lockTokenOrMessage: string | { lockToken: string }, options);
+  defer(lockTokenOrMessage: string | { lockToken: string }, options);
+  deadLetter(lockTokenOrMessage: string | { lockToken: string }, options);
+}
+```
+
+This would match with what other languages, like Java, are doing for their primary flow.
+
+Some issues:
+
+- Method clutter on the `Receiver`.
+- Names have no logical grouping to indicate that they are settlement methods (this problem also exists
```

Comment:
Grouping a few comments into one since they're somewhat interrelated.

I'm torn between "do nothing" and "additional methods on receiver", because that scenario would imply that processing has been so decoupled from receipt that the user might be better off using a different design pattern or tooling (persisted datastore?), or using a map as you say (since they'd need to be persisting message locks _anyway_); do we have real world examples of users who have needed this functionality?

If it becomes necessary I actually feel quite comfortable with additional receiver methods, we can call them "completeMessages" and give the prefixing (suffixing) that you mention, as well as clearer implicature AND the justification that we're providing a novel functionality, settling many events at once, since we can take a list of lock tokens or list of messages.  It's definitely stretching the "only one way to do things" aspiration, but not so far that I feel guilty.

Reply:
settle messages through lock token is an advanced user scenario, right? which means only a small amount of users would use the feature.

I'm thinking of having only one sugar method on the Receiver like:

```
class SettlementType(Enum):
    COMPLETE,
    DEFER,
    DEAD_LETTER,
    ABANDON

receiver.settle_message(lock_token | message, SettlementType.COMPLETE)
``` 
```

Reply:
@KieranBrantnerMagee, re: "do we have real world examples of users who have needed this functionality?"

I'm checking with @ramya-rao-a about that. We didn't have this in track 1 so I think we can at least, from the JS side, assess some level of desire. I'll do an issue search just to be thorough.

@yunhaoling - I like the idea, let's refine a bit:
- Why would we collapse all the settlement methods here but not on the `ReceivedMessageWithLock` itself?
- deadLetter() allows for two named fields (reason and description). How would that signature be exposed in this proposed model?

Reply:
Heh, Richard preempted my question re: one settlement method.  I thought through that initially because I wanted to minimize the surface area for esoteric functionality, as you say Adam, but the realization that we have settlement-type-specific-options throws a wrench in that, and I've not come up with a clever workaround.  (If you can think of one I'm all ears)

Reply:
For now I'm going to go with "do nothing" just to get the ball rolling.